File: python/example/dynamic_nsa_v1.py
#!/usr/bin/env python3
# coding: utf-8
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This file is a part of the CANN Open Software.
# Licensed under CANN Open Software License Agreement Version 1.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ======================================================================================================================
"""
"""
import logging
import math
from dataclasses import dataclass
from typing import List
from enum import Enum
import numpy as np
import pto

# RuntimeConfig KEYS
SG_CYCLE_UPPER_BOUND = "cycle_upper_bound"
L1_REUSE = "l1_reuse"
CUBE_NBUFFER_MAP = "cube_nbuffer_map"
COPYIN_THRESHOLD = "copyin_threshold"
NBUFFER_MERGE_MODE = "nbuffer_merge_mode"
KEY_SUPPORT_DYNAMIC_UNALIGNED = "SUPPORT_DYNAMIC_UNALIGNED"
TILE_VEC_DIMS = 2
TILE_CUBE_DIMS = 6
SHAPE_DIM2 = 2
SHAPE_DIM3 = 3
SHAPE_DIM4 = 4
SHAPE_DIM5 = 5


NUM_65536 = 65536
NUM_1 = 1
NUM_2 = 2
NUM_3 = 3
NUM_4 = 4
NUM_5 = 5
NUM_8 = 8
NUM_16 = 16
NUM_20 = 20
NUM_32 = 32
NUM_64 = 64
NUM_128 = 128
NUM_256 = 256
NUM_512 = 512
NUM_1024 = 1024
NUM_1536 = 1536
NUM_7168 = 7168
NUM_4096 = 4096
DF_1E_5 = 1e-5


class GateMode(Enum):
    standard = 0
    simple = 1


class SimpleParams:
    b: int
    s: int
    s2: int
    d: int
    m: int
    k: int
    n: int
    n2: int
    right: int
    h: int
    q_lora_rank: int
    kv_lora_rank: int
    qk_rope_head_dim: int
    qk_nope_head_dim: int
    q_head_dim: int
    cache_mode: str
    block_size: int
    vec_tile = []
    cube_m_tile = []
    cube_k_tile = []
    cube_n_tile = []
    tile_b: int


    @staticmethod
    def get_common_garams():
        params: SimpleParams = SimpleParams()
        params.n2 = NUM_1
        params.s = NUM_1
        params.h = NUM_7168 # 7168
        params.q_lora_rank = NUM_1536 # 1536
        params.kv_lora_rank = NUM_512 # 512
        params.qk_rope_head_dim = NUM_64 # 64
        params.qk_nope_head_dim = NUM_128 # 128
        params.q_head_dim = params.qk_rope_head_dim + params.qk_nope_head_dim
        params.cache_mode = "BNSD"
        params.block_size = NUM_128 # 128
        return params


    @staticmethod
    def get_low_params():
        params: SimpleParams = SimpleParams.get_common_garams()
        params.b = NUM_4
        params.n = NUM_32
        params.s2 = NUM_256
        return params


    @staticmethod
    def get_high_params():
        params: SimpleParams = SimpleParams.get_common_garams()
        params.b = NUM_32
        params.n = NUM_128
        params.s2 = NUM_4096
        return params


def gen_gated_score(**kwargs):
    x = kwargs.get("x")
    gate_w1 = kwargs.get("gate_w1")
    gate_w2 = kwargs.get("gate_w2")
    gate_sim_w1 = kwargs.get("gate_sim_w1")
    gating_score = kwargs.get("gating_score")
    gate_mode = kwargs.get("gate_mode")

    d_type = x.dtype

    b = x.shape[0]
    s = x.shape[1]
    h = x.shape[2]
    n1 = gate_w2.shape[1] // 3
    tile_b = b
    tile_s = s
    tile_bs = tile_b * tile_s

    b_loop = b // tile_b
    s_loop = s // tile_s
    for b_idx in pto.loop(0, b_loop, 1, name="LOOP_L0_bIdx_gated_score", idx_name="b_idx"):
        def insideb_idx_loop(b_idx):
            for s_idx in pto.loop(0, s_loop, 1, name="LOOP_L0_sIdx_gated_score", idx_name="s_idx"):
                def insides_idx_loop(s_idx):
                    pto.set_vec_tile_shapes(tile_b, tile_s, h)
                    pto.set_cube_tile_shapes([tile_bs, tile_bs], [NUM_128, NUM_128], [NUM_128, NUM_128])
                    b_ofs = b_idx * tile_b
                    s_ofs = s_idx * tile_s
                    bs_ofs = b_ofs * s_ofs

                    x_reshape = pto.reshape(x, [b * s, h])
                    x_view = pto.view(x_reshape, [tile_bs, h], [bs_ofs, 0])
                    mm1_res = pto.matmul(pto.DataType.DT_FP32, x_reshape, gate_w1)

                    pto.set_vec_tile_shapes(1, h)
                    sigmoid_res = pto.sigmoid(mm1_res)
                    sigmoid_res = pto.cast(sigmoid_res, d_type)
                    pto.set_cube_tile_shapes([tile_bs, tile_bs], [NUM_128, NUM_128], [NUM_16, NUM_16])
                    mm2_res = pto.matmul(pto.DataType.DT_FP32, sigmoid_res, gate_w2)
                    pto.set_vec_tile_shapes(tile_bs, n1)

                    res = pto.reshape(mm2_res, [tile_b, tile_s, 3, n1])
                    pto.set_vec_tile_shapes(1, tile_s, 3, n1)

                    res = pto.transpose(res, [2, 3])
                    if gating_score.dtype != pto.DataType.DT_FP32:
                        res = pto.cast(res, d_type)
                    pto.assemble(res, [b_ofs, s_idx, 0, 0], gating_score)
                insides_idx_loop(s_idx)
        insideb_idx_loop(b_idx)


def gen_gated_score_compute(**kwargs):
    x = kwargs.get("x")
    gate_w1 = kwargs.get("gate_w1")
    gate_w2 = kwargs.get("gate_w2")
    gate_sim_w1 = kwargs.get("gate_sim_w1")
    gating_score = kwargs.get("gating_score")
    gate_mode = kwargs.get("gate_mode")
    input_tensors = [x, gate_w1, gate_w2, gate_sim_w1]
    output_tensors = [gating_score]
    with pto.function("fused_compress_kv_select", input_tensors, output_tensors):
        gen_gated_score(
            x=x,
            gate_w1=gate_w1,
            gate_w2=gate_w2,
            gate_sim_w1=gate_sim_w1,
            gating_score=gating_score,
            gate_mode=gate_mode
        )


def test_nsa(params):
    b = params.b
    s = params.s
    n = params.n
    h = params.h

    d_type = pto.DataType.DT_FP16

    x_shape = [b, s, h]
    gate_w1_shape = [h, 4 * h]
    gate_w2_shape = [4 * h, 3 * n]
    gate_sim_w1_shape = [h, 3 * n]
    gating_score_shape = [b, s, n, 3]
    temp_shape = [b * s, n * 3]
    mm1_shape = [b * s, 4 * h]

    x = pto.tensor(x_shape, d_type, "x")
    gate_w1 = pto.tensor(gate_w1_shape, d_type, "gateW1")
    gate_w2 = pto.tensor(gate_w2_shape, d_type, "gateW2")
    gate_sim_w1 = pto.tensor(gate_sim_w1_shape, d_type, "gateSimW1")
    gating_score = pto.tensor(gating_score_shape, d_type, "gatingScore")

    gen_gated_score_compute(
        x=x,
        gate_w1=gate_w1,
        gate_w2=gate_w2,
        gate_sim_w1=gate_sim_w1,
        gating_score=gating_score,
        gate_mode=GateMode.standard,
        )


def main():
    params = SimpleParams.get_high_params()

    params.h = 128
    params.s = NUM_2

    test_nsa(params)


if __name__ == "__main__":
    main()

File: python/example/cpp_reference/src/dynamic_nsa_v1.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This file is a part of the CANN Open Software.
 * Licensed under CANN Open Software License Agreement Version 1.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

/*!
 * \file dynamic_nsa.cpp
 * \brief
 */

#include "interface/function/function.h"
#include "interface/configs/config_manager.h"

using namespace npu::tile_fwk;

constexpr int NUM_65536 = 65536;
constexpr int NUM_128 = 128;
constexpr int NUM_2 = 2;
constexpr int NUM_16 = 16;

namespace npu::tile_fwk {
enum GateMode {
    standard,
    simple
};

struct SimpleParams {
    int b;
    int s;
    int s2;
    int d;
    int m;
    int k;
    int n;
    int n2;
    int right;
    int h;
    int q_lora_rank;
    int kv_lora_rank;
    int qk_rope_head_dim;
    int qk_nope_head_dim;
    int q_head_dim;
    std::string cacheMode;
    int blockSize;
    std::vector<int> vecTile;
    std::vector<int> cubeMTile;
    std::vector<int> cubeKTile;
    std::vector<int> cubeNTile;
    int tileB;
    static SimpleParams getCommonParams() {
        SimpleParams params;
        params.n2 = 1;
        params.s = 1;
        params.h = 7168; // 7168
        params.q_lora_rank = 1536; // 1536
        params.kv_lora_rank = 512; // 512
        params.qk_rope_head_dim = 64; // 64
        params.qk_nope_head_dim = 128; // 128
        params.q_head_dim = params.qk_rope_head_dim + params.qk_nope_head_dim;
        params.cacheMode = "BNSD";
        params.blockSize = 128; // 128
        return params;
    }

    static SimpleParams getLowParams() {
        SimpleParams params = getCommonParams();
        params.b = 4; // 4
        params.n = 32; // 32
        params.s2 = 256; // 256
        return params;
    }

    static SimpleParams getHighParams() {
        SimpleParams params = getCommonParams();
        params.b = 32; // 32
        params.n = 128; // 128
        params.s2 = 4096; // 4096
        return params;
    }
};


void GenGatedScore(const Tensor &x, const Tensor &gateW1, const Tensor &gateW2, const Tensor &gateSimW1,
    Tensor &gatingScore, GateMode gateMode) {
    (void)gateSimW1;
    (void)gateMode;
    DataType dType = x.GetStorage()->Datatype();

    int b = x.GetShape()[0];
    int s = x.GetShape()[1];
    int h = x.GetShape()[2];
    int n1 = gateW2.GetShape()[1] / 3;
    int tileB = b;
    int tileS = s;
    int tileBS = tileB * tileS;

    SymbolicScalar bLoop = b / tileB;
    SymbolicScalar sLoop = s / tileS;
    LOOP("LOOP_L0_bIdx_gated_score", FunctionType::DYNAMIC_LOOP, bIdx, LoopRange(0, bLoop, 1)) {
        LOOP("LOOP_L0_sIdx_gated_score", FunctionType::DYNAMIC_LOOP, sIdx, LoopRange(0, sLoop, 1)) {
            TileShape::Current().SetVecTile({tileB, tileS, h});
            TileShape::Current().SetCubeTile({tileBS, tileBS}, {NUM_128, NUM_128}, {NUM_128, NUM_128});
            SymbolicScalar bOfs = bIdx * tileB;
            SymbolicScalar sOfs = sIdx * tileS;
            SymbolicScalar bsOfs = bOfs * sOfs;

            auto xReshape = Reshape(x, {b * s, h});
            auto xView = View(xReshape, {tileBS, h}, {bsOfs, 0});
            auto mm1Res = Matrix::Matmul(DT_FP32, xReshape, gateW1);

            TileShape::Current().SetVecTile({1, h});
            auto sigmoidRes = Sigmoid(mm1Res);
            sigmoidRes = Cast(sigmoidRes, dType);
            TileShape::Current().SetCubeTile({tileBS, tileBS}, {NUM_128, NUM_128}, {NUM_16, NUM_16});
            auto mm2Res = Matrix::Matmul(DT_FP32, sigmoidRes, gateW2);
            TileShape::Current().SetVecTile({tileBS, n1});

            auto res = Reshape(mm2Res, {tileB, tileS, 3, n1});
            TileShape::Current().SetVecTile({1, tileS, 3, n1});

            res = Transpose(res, {2, 3});
            if (gatingScore.GetStorage()->Datatype() != DT_FP32) {
                res = Cast(res, dType);
            }
            Assemble(res, {bOfs, sIdx, 0, 0}, gatingScore);
        }
    }
}

void GenGatedScoreCompute(const Tensor &x, const Tensor &gateW1, const Tensor &gateW2, const Tensor &gateSimW1,
    Tensor &gatingScore, GateMode gateMode) {
    FUNCTION("FusedCompressKvSelect", {x, gateW1, gateW2, gateSimW1}, {gatingScore}) {
        GenGatedScore(x, gateW1, gateW2, gateSimW1, gatingScore, gateMode);
    }
}


template <typename T = npu::tile_fwk::float16, typename wDtype = int8_t, bool splitK = false, bool nz = true,
    bool isSmooth = true, bool usePrefetch = true>
void TestNsa(const SimpleParams &params) {
    int b = params.b;
    int s = params.s;
    int n = params.n;
    int h = params.h;

    DataType dType = (std::is_same<T, npu::tile_fwk::float16>::value) ? DT_FP16 : DT_BF16;

    std::vector<int64_t> x_shape = {b, s, h};
    std::vector<int64_t> gateW1Shape = {h, 4 * h};
    std::vector<int64_t> gateW2Shape = {4 * h, 3 * n};
    std::vector<int64_t> gateSimW1Shape = {h, 3 * n};
    //    std::vector<int64_t> gatingScoreShape = {b, n, s, 3};
    std::vector<int64_t> gatingScoreShape = {b, s, n, 3};
    std::vector<int64_t> tempShape = {b * s, n * 3};
    std::vector<int64_t> mm1Shape = {b * s, 4 * h};

    Tensor x(dType, x_shape, "x");
    Tensor gateW1(dType, gateW1Shape, "gateW1");
    Tensor gateW2(dType, gateW2Shape, "gateW2");
    Tensor gateSimW1(dType, gateSimW1Shape, "gateSimW1");
    Tensor gatingScore(dType, gatingScoreShape, "gatingScore");
    GenGatedScoreCompute(x, gateW1, gateW2, gateSimW1, gatingScore, GateMode::standard);
}

} // namespace npu::tile_fwk


int main() {
    SimpleParams params = SimpleParams::getHighParams();
    params.h = NUM_128;
    params.s = NUM_2;
    TestNsa<npu::tile_fwk::float16>(params);
}